src/System.Linq.Expressions/src/System/Runtime/CompilerServices/ReadOnlyCollectionBuilder.cs

```diff
@@ -462,15 +526,18 @@ public ReadOnlyCollection<T> ToReadOnlyCollection()
             return new TrueReadOnlyCollection<T>(items);
         }
 
+        // Ensures that the capacity of this list is at least the given minimum
+        // value. If the currect capacity of the list is less than min, the
+        // capacity is increased to twice the current capacity or to min,
+        // whichever is larger.
         private void EnsureCapacity(int min)
         {
             if (_items.Length < min)
             {
-                int newCapacity = DefaultCapacity;
-                if (_items.Length > 0)
-                {
-                    newCapacity = _items.Length * 2;
-                }
+                int newCapacity = _items.Length > 0 ? _items.Length * 2 : DefaultCapacity;
+                // Allow the list to grow to maximum possible capacity (~2G elements) before encountering overflow.
+                // Note that this check works even when _items.Length overflowed thanks to the (uint) cast
+                if ((uint)newCapacity > int.MaxValue) newCapacity = int.MaxValue;
```

Comment:
When doing the trick where you cast to `uint`, you need to do so for both sides, e.g: `((uint)newCapacity > (uint)int.MaxValue)`. The reason is if you attempt to compare an `int` and a `uint`, the compiler will convert both sides to `long` before comparing. So what you currently have is equivalent to

```cs
if ((long)(uint)newCapacity > (long)int.MaxValue) newCapacity = int.MaxValue;
```

which is semantically correct, but is slower.